src/JanusStream.cpp

```diff
@@ -0,0 +1,116 @@
+/**
+ * @file JanusStream.h
+ * @author Daniel Stiner (danstiner@gmail.com)
+ * @date 2021-03-10
+ * @copyright Copyright (c) 2021 Daniel Stiner
+ */
+
+#include "JanusStream.h"
+
+#pragma region Constructor/Destructor
+JanusStream::JanusStream(
+    ftl_channel_id_t channelId,
+    ftl_stream_id_t streamId,
+    MediaMetadata mediaMetadata) :
+    channelId(channelId),
+    streamId(streamId),
+    mediaMetadata(mediaMetadata)
+{ }
+#pragma endregion
+
+#pragma region Public methods
+void JanusStream::SendRtpPacket(const std::vector<std::byte>& packet)
+{
+    std::lock_guard lock(mutex);
+
+    for (const auto& session : viewerSessions)
+    {
+        session->SendRtpPacket(packet, mediaMetadata);
+    }
+
+    for (const auto& relay : relays)
+    {
+        relay.Client->RelayPacket(packet);
+    }
+}
+
+void JanusStream::AddViewerSession(JanusSession* session)
+{
+    std::lock_guard lock(mutex);
+    viewerSessions.insert(session);
+}
+
+size_t JanusStream::RemoveViewerSession(JanusSession* session)
+{
+    std::lock_guard lock(mutex);
+    return viewerSessions.erase(session);
+}
+
+std::unordered_set<JanusSession*> JanusStream::RemoveAllViewerSessions()
+{
+    std::lock_guard lock(mutex);
+    std::unordered_set<JanusSession*> viewerSessions;
+    viewerSessions.swap(viewerSessions);
+    return viewerSessions;
+}
+
+size_t JanusStream::GetViewerCount() const
+{
+    std::lock_guard lock(mutex);
+    return viewerSessions.size();
+}
+
+void JanusStream::AddRelayClient(const std::string targetHostname, std::unique_ptr<FtlClient> client)
+{
+    std::lock_guard lock(mutex);
+    relays.push_back(Relay { .TargetHostname = targetHostname, .Client = std::move(client) });
+}
+
+
+size_t JanusStream::StopRelay(const std::string& targetHostname)
+{
+    std::lock_guard lock(mutex);
+    size_t numRelaysRemoved = 0;
+    for (auto it = relays.begin();
+        it != relays.end();)
+    {
+        Relay& relay = *it;
+        if (relay.TargetHostname == targetHostname) {
+            relay.Client->Stop();
```

Comment:
I worry about holding a lock while we wait for an FtlClient to stop

Reply:
Oh shit you're right, any suggestions? Is it safe to just unlock during this?